### PR TITLE
FEATURE: Scroll to first message when clicking date in chat

### PR DESCRIPTION
**Pull request by jancernik**
This PR adds a new parameter to fetch chat messages. `target_date` can be used to fetch messages by a specific date string. Note that it does not need to be the `created_at` date of an existing message, it can be any date.
Similar to `target_message_id`, it retrieves an array of past and future messages following the query limits.

### CODE REVIEW

File: plugins/chat/app/controllers/chat/api/channels_controller.rb [hunk 1]
```diff
@@ -68,21 +68,23 @@ def create
   end
 
   def show
-    if params[:target_message_id].present? || params[:include_messages].present? ||
-         params[:fetch_from_last_read].present?
+    if params[:target_message_id].present? || params[:target_date].present? ||
```
Comment (pmusaraj) [inline plugins/chat/app/controllers/chat/api/channels_controller.rb:71]: Since this adds a new parameter to the controller, it would be good to describe how to use this, maybe in the PR description.

**Review by pmusaraj — COMMENTED** [review]
Looks generally good to me, but @jjaffeux and @martin-brennan know the chat context much better than me.

File: plugins/chat/assets/javascripts/discourse/components/chat-message-separator-date.hbs [hunk 2]
```diff
@@ -10,7 +10,16 @@
       {{chat/track-message-separator-date}}
     >
       <span class="chat-message-separator__text">
-        <span>{{@message.firstMessageOfTheDayAt}}</span>
+        <DButton
+          type="button"
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-message-separator-date.hbs:14]: is this needed? I suspect our DButton does it by default

File: plugins/chat/assets/javascripts/discourse/components/chat-message-separator-date.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+import Component from "@glimmer/component";
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-message-separator-date.js:1]: is this file necessary?

File: plugins/chat/assets/javascripts/discourse/components/chat-message-separator-date.js
[the remark this answers is not in the document]
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/components/chat-message-separator-date.js:1]: ah lol I did the same comment without reload the page :p

**Comment by jjaffeux** [comment]
@jancernik I think there's an error with channels when threads are enabled.

You can enable threads like this:

```
SiteSetting.enable_experimental_chat_threaded_discussions = true
Chat::Channel.find(__CHANNEL_ID__).update!(thread_enabled: true)
```

This is the error I see at the moment when click on the date of a channel with threading enabled:

```
REQUEST URL 500 while getting:
http://127.0.0.1:4200/chat/api/channels/2?page_size=50&include_messages=true&target_date=2023-06-09T00%3A00%3A00%2B02%3A00

NoMethodError in Chat::Api::ChannelsController#show

undefined method `thread_reply?' for nil:NilClass
Extracted source (around line #141):
```

File: plugins/chat/app/queries/chat/messages_query.rb [hunk 4]
```diff
@@ -3,10 +3,10 @@
 module Chat
   # Queries messages for a specific channel. This can be used in two modes:
   #
-  # 1. Query messages around a target_message_id. This is used when a user
-  #    needs to jump to the middle of a messages stream or load around their
-  #    last read message ID. There is no pagination or direction here, just
-  #    a limit on past and future messages.
+  # 1. Query messages around a target_message_id or target_date. This is used
+  #    when a user needs to jump to the middle of a messages stream or load
+  #    around their last read message ID. There is no pagination or direction
```
Comment (jjaffeux) [inline plugins/chat/app/queries/chat/messages_query.rb:8]: Let's talk of target after, so we don't have to repeat message ID or date each time

```suggestion
  #    around a target. There is no pagination or direction
```

File: plugins/chat/app/queries/chat/messages_query.rb [hunk 5]
```diff
@@ -28,6 +28,7 @@ class MessagesQuery
     # @param thread_id [Integer] (optional) The thread ID to filter messages by.
     # @param target_message_id [Integer] (optional) The message ID to query around.
     #   It is assumed that the caller already checked if this exists.
+    # @param target_date [String] (optional) The date to query around.
```
Comment (martin-brennan) [inline plugins/chat/app/queries/chat/messages_query.rb:31]: Does the date need to be in a particular format? If so we need to probably put this in the comment here / check for this in the query

File: plugins/chat/app/queries/chat/messages_query.rb [hunk 6]
```diff
@@ -151,5 +157,32 @@ def self.query_paginated_messages(
         can_load_more_future: can_load_more_future,
       }
     end
+
+    def self.query_by_date(target_date, channel, messages)
+      past_messages =
+        messages
+          .where("created_at < ?", target_date)
```
Comment (martin-brennan) [inline plugins/chat/app/queries/chat/messages_query.rb:164]: Would do `<=` for this and the future ones, that way we can be sure we don't randomly miss 1 message because it miraculously landed exactly on the datetime

File: plugins/chat/app/services/chat/channel_view_builder.rb [hunk 7]
```diff
@@ -50,6 +50,7 @@ class Contract
       attribute :direction, :string # (optional)
       attribute :page_size, :integer # (optional)
       attribute :fetch_from_last_read, :boolean # (optional)
+      attribute :target_date, :string # (optional)
```
Comment (martin-brennan) [inline plugins/chat/app/services/chat/channel_view_builder.rb:53]: Actually this service is a better place to validate that this is a correct date/time -- we need to add a `validate target_date_ok?` method to validate it. It's the same as rails model validations e.g. 

https://github.com/discourse/discourse/blob/a4630d220b55c45b9b72cbc1e598c0d626f2e598/app/models/invite.rb#L39-L39

https://github.com/discourse/discourse/blob/a4630d220b55c45b9b72cbc1e598c0d626f2e598/app/models/invite.rb#L70-L72

**Review by martin-brennan — COMMENTED** [review]
This is looking good so far just a couple of minor things to fix :+1:

File: plugins/chat/app/services/chat/channel_view_builder.rb (hunk 7 above)
Comment (jjaffeux) [reply plugins/chat/app/services/chat/channel_view_builder.rb:54]: I think date is very hard to validate because it can be a string or anything with can be casted to string. What validation would you write?

Comment (martin-brennan) [reply plugins/chat/app/services/chat/channel_view_builder.rb:54]: Hmm I guess you are right, I had assumed we would have a pattern for this. Fair enough to leave it out.

**Review by jjaffeux — APPROVED** [review]
